test/Unit/Durable/ActivityInvocationTrackerTests.cs

```diff
@@ -20,101 +21,209 @@ public class ActivityInvocationTrackerTests
+                CreateHistory("FunctionB", scheduled: true, completed: true, output: "\"Result2\""),
+                CreateHistory("FunctionA", scheduled: true, completed: true, output: "\"Result3\"")
+            );
+            var orchestrationContext = new OrchestrationContext { History = history };
+            var allOutput = new List<object>();
 
-            stopwatch.Start();
-            activityInvocationTracker.ReplayActivityOrStop(FunctionName, FunctionInput, orchestrationContext, _ => { });
-            stopwatch.Stop();
+            var activityInvocationTracker = new ActivityInvocationTracker();
 
-            // Check if ReplayActivityOrStop was actually blocked
-            if (expectedWaitForStop)
+            // Replay FunctionA only
+            for (var i = 0; i < 2; ++i)
             {
-                Assert.True(stopwatch.ElapsedMilliseconds > delayBeforeStopping.TotalMilliseconds * 0.8);
+                activityInvocationTracker.ReplayActivityOrStop(
+                    "FunctionA", FunctionInput, orchestrationContext, noWait: false,
+                    output => { allOutput.Add(output); });
             }
-            else
+
+            // Expect FunctionA results only
+            Assert.Equal(new[] { "Result1", "Result3" }, allOutput);
+        }
+
+        [Theory]
+        [InlineData(false, false)]
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        [InlineData(true, true)]
+        public void ReplayActivityOrStop_OutputsActivityTask_WhenNoWaitRequested(
+            bool scheduled, bool completed)
+        {
+            var history = CreateHistory(scheduled: scheduled, completed: completed, output: InvocationResultJson);
+            var orchestrationContext = new OrchestrationContext { History = history };
+            var allOutput = new List<ActivityInvocationTask>();
+
+            var activityInvocationTracker = new ActivityInvocationTracker();
+            activityInvocationTracker.ReplayActivityOrStop(
+                FunctionName, FunctionInput, orchestrationContext, noWait: true,
+                output => { allOutput.Add((ActivityInvocationTask)output); });
+
+            VerifyCallActivityActionAdded(orchestrationContext);
+            Assert.Equal(FunctionName, allOutput.Single().Name);
+        }
+
+        [Theory]
+        [InlineData(true, true)]
+        public void WaitForActivityTasks_OutputsActivityResults_WhenAllTasksCompleted(
+            bool scheduled, bool completed)
+        {
+            var history = new[]
             {
-                Assert.True(stopwatch.ElapsedMilliseconds < delayBeforeStopping.TotalMilliseconds * 0.2);
-            }
+                CreateHistory("FunctionA", scheduled: scheduled, completed: completed, output: "\"Result1\""),
+                CreateHistory("FunctionA", scheduled: scheduled, completed: completed, output: "\"Result2\""),
```

Comment:
In the `CreateHistory` we have two `FunctionA`. What's the purpose behind this? I apologize if this is supposed to be obvious. I see that below we check the outputs order which should be "Result1", "Result2", "Result3" which makes sense to me.

Reply:
This is actually important, but the purpose is not immediately obvious, you are right. I'm adding a comment.

Reply:
Great, thank you!